### DISCUSSION TITLE
Are database migrations expected to be cross-compatible?

### DISCUSSION THREAD

**Discussion by Sillyvan** [loco-rs/loco#865] [Q&A] [answered]
Title should explain the question. I ran into multiple people having issues with this in the last couple of days.
In this case caused by choosing a Starter choosing postgres and then manually switch to sqlite.

Maybe we havent caught all differences but the one me and the community saw were related to enums. Basically in postgres we generate and drop enum types. in sqlite we just have text (or int). Both end up mapping to actual rust enums.

![image](https://github.com/user-attachments/assets/358cc1ee-c17e-476c-b9a9-b5b13abfca43)

I mentioned it here https://github.com/loco-rs/loco/discussions/836 im sadly not familiar with SeaORM at all so i hacked something together.

**Comment by jondot** [comment, marked as answer]
I believe for the extent this is possible we would always want compatibility. Which may mean in this case to find a way to abstract enums and not use a native type. 

That said, this direction is holding a line of “performance is good enough” but we should be able to let people use native types when they need, since almost always it will be the better option (people will rarely switch database providers after they went to production). 


in other words if it is easy we should go for it but remember that mostly it is a development convenience. On most cases people will just use Postgres in production and be OK with native types.

**Reply by Sillyvan** [reply]
Totally agree! But i think we should make sure that at least the [examples](https://github.com/loco-rs/loco/tree/master/examples) have migrations that work with both. Makes it easier to test things.

**Reply by Sillyvan** [reply]
It should also be clear that the classic "Sqlite for local, postgres for prod" is not as it sounds. By using sqlite locally you will end up losing native features on production currently

**Reply by jondot** [reply]
Correct. Fixing up the demo in example is something we should do — right just removing the roles implementation is what we considered in order to keep it simple but still be a kitchen sink kind of thing.

**Reply by Sillyvan** [reply]
Thats nice. Im just trying to document some issues me (and others) get confused about while learning how the entire code base works. Hopefully i can move to committing more discussing less soon!

**Reply by jondot** [reply]
Thanks 🙏 
Really appreciate the efforts! It makes this project more fun ❤️
